src/main/java/picard/analysis/AlignmentSummaryMetricsCollector.java

```diff
@@ -103,148 +104,206 @@ public GroupAlignmentSummaryMetricsPerUnitMetricCollector(final String sample, f
-                        if (badCyclePercentage >= 0.8) {
-                            metrics.BAD_CYCLES++;
-                        }
-                    }
+            collectReadData(record);
+            collectQualityData(record, ref);
+        }
 
-                    if(doRefMetrics) {
-                        metrics.PCT_PF_READS_ALIGNED =  MathUtil.divide((double) metrics.PF_READS_ALIGNED, (double) metrics.PF_READS);
-                        metrics.PCT_READS_ALIGNED_IN_PAIRS =  MathUtil.divide((double) metrics.READS_ALIGNED_IN_PAIRS, (double) metrics.PF_READS_ALIGNED);
-                        metrics.PCT_PF_READS_IMPROPER_PAIRS =  MathUtil.divide((double) metrics.PF_READS_IMPROPER_PAIRS, (double) metrics.PF_READS_ALIGNED);
-                        metrics.STRAND_BALANCE =  MathUtil.divide(numPositiveStrand, (double) metrics.PF_READS_ALIGNED);
-                        metrics.PCT_CHIMERAS =  MathUtil.divide(this.chimeras, (double) this.chimerasDenominator);
-                        metrics.PF_INDEL_RATE =  MathUtil.divide(this.indels, (double) metrics.PF_ALIGNED_BASES);
-                        metrics.PF_MISMATCH_RATE =  MathUtil.divide(mismatchHistogram.getSum(), (double) nonBisulfiteAlignedBases);
-                        metrics.PF_HQ_ERROR_RATE =  MathUtil.divide(hqMismatchHistogram.getSum(), (double) hqNonBisulfiteAlignedBases);
-
-                        metrics.PF_HQ_MEDIAN_MISMATCHES = hqMismatchHistogram.getMedian();
+        @Override
+        public void finish() {
+            //summarize read data
+            if (metrics.TOTAL_READS > 0) {
+                metrics.PCT_PF_READS = (double) metrics.PF_READS / (double) metrics.TOTAL_READS;
+                metrics.PCT_ADAPTER = this.adapterReads / (double) metrics.PF_READS;
+                metrics.MEAN_READ_LENGTH = readLengthHistogram.getMean();
+
+                //Calculate BAD_CYCLES
+                metrics.BAD_CYCLES = 0;
+                for (final Histogram.Bin<Integer> cycleBin : badCycleHistogram.values()) {
+                    final double badCyclePercentage = cycleBin.getValue() / metrics.TOTAL_READS;
+                    if (badCyclePercentage >= 0.8) {
+                        metrics.BAD_CYCLES++;
                     }
                 }
+
+                if (doRefMetrics) {
+                    metrics.PCT_PF_READS_ALIGNED = MathUtil.divide((double) metrics.PF_READS_ALIGNED, (double) metrics.PF_READS);
+                    metrics.PCT_READS_ALIGNED_IN_PAIRS = MathUtil.divide((double) metrics.READS_ALIGNED_IN_PAIRS, (double) metrics.PF_READS_ALIGNED);
+                    metrics.PCT_PF_READS_IMPROPER_PAIRS = MathUtil.divide((double) metrics.PF_READS_IMPROPER_PAIRS, (double) metrics.PF_READS_ALIGNED);
+                    metrics.STRAND_BALANCE = MathUtil.divide(numPositiveStrand, (double) metrics.PF_READS_ALIGNED);
+                    metrics.PCT_CHIMERAS = MathUtil.divide(this.chimeras, (double) this.chimerasDenominator);
```

Comment:
I wouldn't normally comment on `this.` but the code is inconsistent in its use. Why do these use `this.` but no other fields do? I think it would be better to omit them completely unless necessary for disambiguation purposes.